File: public/dsa/searching/next_greater_element_in_sorted_array/condition_based/next_greater_element_in_sorted_array.py
from typing import List


class Solution:
    def getNextGreaterElement(self, arr: List[int], key: int) -> int:
        n = len(arr)
        if key >= arr[n-1]:
            return key
        l = 0
        r = n - 1
        while l < r:
            m = (l + r) // 2
            if arr[m] <= key:
                l = m + 1
            else:
                r = m
        return arr[l]


# tests-start
arr = [1, 2, 3, 3, 4, 4, 8, 10]
key = 4
sol = Solution()
print(sol.getNextGreaterElement(arr, key))
# tests-end

'''output
8
'''

File: public/dsa/searching/next_greater_element_in_sorted_array/condition_based/next_greater_element_in_sorted_array.cpp
#include <bits/stdc++.h>
using namespace std;

int getNextGreaterElement(vector<int> &arr, int key) {
    // add your logic here
    int n = arr.size();
    if (key >= arr[n - 1]) {
        return key;
    }
    int l = 0;
    int r = n - 1;
    while (l < r) {
        int m = (l + r) / 2;
        if (arr[m] <= key) {
            l = m + 1;
        } else {
            r = m;
        }
    }
    return arr[l];
}

// tests-start
int main() {
    vector<int> arr = {1, 2, 3, 3, 4, 4, 8, 10};
    int key = 4;
    cout << getNextGreaterElement(arr, key) << endl;
    return 0;
}
// tests-end

/*output
8
*/
